File: src/rules/cli-commands.ts
import type { Finding } from "../types";
import { makeFinding } from "../util";

/**
 * Table of CLI command migrations. Patterns are applied in order of declaration.
 * Each `match` is a regex with a single capture that will be replaced by `replace`
 * (supports $1..$n).
 */
interface Mapping {
  match: RegExp;
  replace: string;
  ruleId: string;
  message: string;
}

const MAPPINGS: Mapping[] = [
  // npm ---------------------------------------------------------------
  { match: /\bnpm\s+install\s+--save-dev\b/g,  replace: "bun add --dev",            ruleId: "cli/npm-install-dev",   message: "npm install --save-dev → bun add --dev" },
  { match: /\bnpm\s+install\s+-D\b/g,          replace: "bun add --dev",            ruleId: "cli/npm-install-D",     message: "npm install -D → bun add --dev" },
  { match: /\bnpm\s+install\s+--save\b/g,      replace: "bun add",                  ruleId: "cli/npm-install-save",  message: "npm install --save → bun add" },
  { match: /\bnpm\s+i\s+--save-dev\b/g,        replace: "bun add --dev",            ruleId: "cli/npm-i-dev",         message: "npm i --save-dev → bun add --dev" },
  { match: /\bnpm\s+i\s+-D\b/g,                replace: "bun add --dev",            ruleId: "cli/npm-i-D",           message: "npm i -D → bun add --dev" },
  { match: /\bnpm\s+install\b(?!\s+-g)/g,      replace: "bun install",              ruleId: "cli/npm-install",       message: "npm install → bun install" },
  { match: /\bnpm\s+i\b(?!\s+-g)/g,            replace: "bun install",              ruleId: "cli/npm-i",             message: "npm i → bun install" },
  { match: /\bnpm\s+install\s+-g\b/g,          replace: "bun install -g",           ruleId: "cli/npm-install-global",message: "npm install -g → bun install -g" },
  { match: /\bnpm\s+ci\b/g,                    replace: "bun install --frozen-lockfile", ruleId: "cli/npm-ci",      message: "npm ci → bun install --frozen-lockfile" },
  { match: /\bnpm\s+run\s+/g,                  replace: "bun run ",                 ruleId: "cli/npm-run",           message: "npm run → bun run" },
  { match: /\bnpm\s+test\b/g,                  replace: "bun test",                 ruleId: "cli/npm-test",          message: "npm test → bun test" },
  { match: /\bnpm\s+start\b/g,                 replace: "bun start",                ruleId: "cli/npm-start",         message: "npm start → bun start" },
  { match: /\bnpm\s+uninstall\b/g,             replace: "bun remove",               ruleId: "cli/npm-uninstall",     message: "npm uninstall → bun remove" },
  { match: /\bnpm\s+rm\b/g,                    replace: "bun remove",               ruleId: "cli/npm-rm",            message: "npm rm → bun remove" },
  { match: /\bnpm\s+update\b/g,                replace: "bun update",               ruleId: "cli/npm-update",        message: "npm update → bun update" },
  { match: /\bnpm\s+outdated\b/g,              replace: "bun outdated",             ruleId: "cli/npm-outdated",      message: "npm outdated → bun outdated" },
  { match: /\bnpm\s+pack\b/g,                  replace: "bun pm pack",              ruleId: "cli/npm-pack",          message: "npm pack → bun pm pack" },
  { match: /\bnpm\s+publish\b/g,               replace: "bun publish",              ruleId: "cli/npm-publish",       message: "npm publish → bun publish" },
  { match: /\bnpm\s+link\b/g,                  replace: "bun link",                 ruleId: "cli/npm-link",          message: "npm link → bun link" },
  { match: /\bnpm\s+init\s+-y\b/g,             replace: "bun init -y",              ruleId: "cli/npm-init-y",        message: "npm init -y → bun init -y" },
  { match: /\bnpm\s+init\b/g,                  replace: "bun init",                 ruleId: "cli/npm-init",          message: "npm init → bun init" },
  { match: /\bnpm\s+exec\s+/g,                 replace: "bunx ",                    ruleId: "cli/npm-exec",          message: "npm exec → bunx" },
  { match: /\bnpx\s+/g,                        replace: "bunx ",                    ruleId: "cli/npx",               message: "npx → bunx" },

  // pnpm --------------------------------------------------------------
  { match: /\bpnpm\s+install\s+--save-dev\b/g, replace: "bun add --dev",            ruleId: "cli/pnpm-install-dev",  message: "pnpm install --save-dev → bun add --dev" },
  { match: /\bpnpm\s+add\s+--save-dev\b/g,     replace: "bun add --dev",            ruleId: "cli/pnpm-add-dev",      message: "pnpm add --save-dev → bun add --dev" },
  { match: /\bpnpm\s+add\s+-D\b/g,             replace: "bun add --dev",            ruleId: "cli/pnpm-add-D",        message: "pnpm add -D → bun add --dev" },
  { match: /\bpnpm\s+install\b/g,              replace: "bun install",              ruleId: "cli/pnpm-install",      message: "pnpm install → bun install" },
  { match: /\bpnpm\s+i\b/g,                    replace: "bun install",              ruleId: "cli/pnpm-i",            message: "pnpm i → bun install" },
  { match: /\bpnpm\s+add\b/g,                  replace: "bun add",                  ruleId: "cli/pnpm-add",          message: "pnpm add → bun add" },
  { match: /\bpnpm\s+remove\b/g,               replace: "bun remove",               ruleId: "cli/pnpm-remove",       message: "pnpm remove → bun remove" },
  { match: /\bpnpm\s+rm\b/g,                   replace: "bun remove",               ruleId: "cli/pnpm-rm",           message: "pnpm rm → bun remove" },
  { match: /\bpnpm\s+run\s+/g,                 replace: "bun run ",                 ruleId: "cli/pnpm-run",          message: "pnpm run → bun run" },
  { match: /\bpnpm\s+test\b/g,                 replace: "bun test",                 ruleId: "cli/pnpm-test",         message: "pnpm test → bun test" },
  { match: /\bpnpm\s+start\b/g,                replace: "bun start",                ruleId: "cli/pnpm-start",        message: "pnpm start → bun start" },
  { match: /\bpnpm\s+update\b/g,               replace: "bun update",               ruleId: "cli/pnpm-update",       message: "pnpm update → bun update" },
  { match: /\bpnpm\s+dlx\s+/g,                 replace: "bunx ",                    ruleId: "cli/pnpm-dlx",          message: "pnpm dlx → bunx" },
  { match: /\bpnpm\s+exec\s+/g,                replace: "bunx ",                    ruleId: "cli/pnpm-exec",         message: "pnpm exec → bunx" },

  // yarn --------------------------------------------------------------
  { match: /\byarn\s+install\s+--frozen-lockfile\b/g, replace: "bun install --frozen-lockfile", ruleId: "cli/yarn-install-frozen", message: "yarn install --frozen-lockfile → bun install --frozen-lockfile" },
  { match: /\byarn\s+add\s+--dev\b/g,          replace: "bun add --dev",            ruleId: "cli/yarn-add-dev",      message: "yarn add --dev → bun add --dev" },
  { match: /\byarn\s+add\s+-D\b/g,             replace: "bun add --dev",            ruleId: "cli/yarn-add-D",        message: "yarn add -D → bun add --dev" },
  { match: /\byarn\s+add\b/g,                  replace: "bun add",                  ruleId: "cli/yarn-add",          message: "yarn add → bun add" },
  { match: /\byarn\s+remove\b/g,               replace: "bun remove",               ruleId: "cli/yarn-remove",       message: "yarn remove → bun remove" },
  { match: /\byarn\s+install\b/g,              replace: "bun install",              ruleId: "cli/yarn-install",      message: "yarn install → bun install" },
  { match: /\byarn\s+run\s+/g,                 replace: "bun run ",                 ruleId: "cli/yarn-run",          message: "yarn run → bun run" },
  { match: /\byarn\s+test\b/g,                 replace: "bun test",                 ruleId: "cli/yarn-test",         message: "yarn test → bun test" },
  { match: /\byarn\s+start\b/g,                replace: "bun start",                ruleId: "cli/yarn-start",        message: "yarn start → bun start" },
  { match: /\byarn\s+dlx\s+/g,                 replace: "bunx ",                    ruleId: "cli/yarn-dlx",          message: "yarn dlx → bunx" },
  // Bare `yarn` → `bun install` (classic yarn treats it that way).
  { match: /^(\s*)yarn(\s|$)/gm,               replace: "$1bun install$2",          ruleId: "cli/yarn-bare",         message: "bare `yarn` → bun install" },
];

/** Ignore lines that start with `#` as shell comments, or are inside the
 *  line `"name": "npm"` of a package.json. We do not have a real parser here
 *  but we want to avoid the obvious false positives. */
const COMMENT_PREFIX = /^\s*(#|\/\/)/;

export function applyCliRules(path: string, source: string): { findings: Finding[]; content: string } {
  const findings: Finding[] = [];
  let out = source;

  for (const rule of MAPPINGS) {
    // find all matches first (positions are for the pre-replace buffer)
    rule.match.lastIndex = 0;
    const hits: { index: number; text: string }[] = [];
    let m: RegExpExecArray | null;
    while ((m = rule.match.exec(out)) !== null) {
      // skip comments in shell/yaml-like files
      const lineStart = out.lastIndexOf("\n", m.index) + 1;
      const lineEnd = out.indexOf("\n", m.index);
      const line = out.slice(lineStart, lineEnd === -1 ? undefined : lineEnd);
      if (COMMENT_PREFIX.test(line)) continue;
      hits.push({ index: m.index, text: m[0] });
    }
    if (hits.length === 0) continue;

    for (const hit of hits) {
      findings.push(
        makeFinding(path, out, hit.index, rule.ruleId, rule.message, hit.text, hit.text.replace(rule.match, rule.replace), { autofix: true }),
      );
    }
    // Apply replacement globally after recording findings.
    rule.match.lastIndex = 0;
    out = out.replace(rule.match, rule.replace);
  }

  return { findings, content: out };
}

File: crates/n2b-core/src/rules/cli_commands.rs
use crate::types::{Finding, MakeFindingOpts};
use crate::util::{line_offsets, make_finding};
use once_cell::sync::Lazy;
use regex::Regex;

struct Mapping {
    re: Regex,
    replace: &'static str,
    rule_id: &'static str,
    message: &'static str,
}

macro_rules! m {
    ($pat:expr, $repl:expr, $id:expr, $msg:expr) => {
        Mapping {
            re: Regex::new($pat).unwrap_or_else(|e| {
                panic!(
                    "invariant: cli_commands mapping pattern '{}' is invalid: {}",
                    $pat, e
                )
            }),
            replace: $repl,
            rule_id: $id,
            message: $msg,
        }
    };
}

static MAPPINGS: Lazy<Vec<Mapping>> = Lazy::new(|| {
    vec![
        // npm
        m!(
            r"\bnpm\s+install\s+--save-dev\b",
            "bun add --dev",
            "cli/npm-install-dev",
            "npm install --save-dev → bun add --dev"
        ),
        m!(
            r"\bnpm\s+install\s+-D\b",
            "bun add --dev",
            "cli/npm-install-D",
            "npm install -D → bun add --dev"
        ),
        m!(
            r"\bnpm\s+install\s+--save\b",
            "bun add",
            "cli/npm-install-save",
            "npm install --save → bun add"
        ),
        m!(
            r"\bnpm\s+i\s+--save-dev\b",
            "bun add --dev",
            "cli/npm-i-dev",
            "npm i --save-dev → bun add --dev"
        ),
        m!(
            r"\bnpm\s+i\s+-D\b",
            "bun add --dev",
            "cli/npm-i-D",
            "npm i -D → bun add --dev"
        ),
        m!(
            r"\bnpm\s+install\b(?:\s+-g)?",
            "bun install",
            "cli/npm-install",
            "npm install → bun install"
        ),
        m!(
            r"\bnpm\s+i\b",
            "bun install",
            "cli/npm-i",
            "npm i → bun install"
        ),
        m!(
            r"\bnpm\s+ci\b",
            "bun install --frozen-lockfile",
            "cli/npm-ci",
            "npm ci → bun install --frozen-lockfile"
        ),
        m!(
            r"\bnpm\s+run\s+",
            "bun run ",
            "cli/npm-run",
            "npm run → bun run"
        ),
        m!(
            r"\bnpm\s+test\b",
            "bun test",
            "cli/npm-test",
            "npm test → bun test"
        ),
        m!(
            r"\bnpm\s+start\b",
            "bun start",
            "cli/npm-start",
            "npm start → bun start"
        ),
        m!(
            r"\bnpm\s+uninstall\b",
            "bun remove",
            "cli/npm-uninstall",
            "npm uninstall → bun remove"
        ),
        m!(
            r"\bnpm\s+rm\b",
            "bun remove",
            "cli/npm-rm",
            "npm rm → bun remove"
        ),
        m!(
            r"\bnpm\s+update\b",
            "bun update",
            "cli/npm-update",
            "npm update → bun update"
        ),
        m!(
            r"\bnpm\s+outdated\b",
            "bun outdated",
            "cli/npm-outdated",
            "npm outdated → bun outdated"
        ),
        m!(
            r"\bnpm\s+pack\b",
            "bun pm pack",
            "cli/npm-pack",
            "npm pack → bun pm pack"
        ),
        m!(
            r"\bnpm\s+publish\b",
            "bun publish",
            "cli/npm-publish",
            "npm publish → bun publish"
        ),
        m!(
            r"\bnpm\s+link\b",
            "bun link",
            "cli/npm-link",
            "npm link → bun link"
        ),
        m!(
            r"\bnpm\s+init\s+-y\b",
            "bun init -y",
            "cli/npm-init-y",
            "npm init -y → bun init -y"
        ),
        m!(
            r"\bnpm\s+init\b",
            "bun init",
            "cli/npm-init",
            "npm init → bun init"
        ),
        m!(
            r"\bnpm\s+exec\s+",
            "bunx ",
            "cli/npm-exec",
            "npm exec → bunx"
        ),
        m!(r"\bnpx\s+", "bunx ", "cli/npx", "npx → bunx"),
        // pnpm
        m!(
            r"\bpnpm\s+install\s+--save-dev\b",
            "bun add --dev",
            "cli/pnpm-install-dev",
            "pnpm install --save-dev → bun add --dev"
        ),
        m!(
            r"\bpnpm\s+add\s+--save-dev\b",
            "bun add --dev",
            "cli/pnpm-add-dev",
            "pnpm add --save-dev → bun add --dev"
        ),
        m!(
            r"\bpnpm\s+add\s+-D\b",
            "bun add --dev",
            "cli/pnpm-add-D",
            "pnpm add -D → bun add --dev"
        ),
        m!(
            r"\bpnpm\s+install\b",
            "bun install",
            "cli/pnpm-install",
            "pnpm install → bun install"
        ),
        m!(
            r"\bpnpm\s+i\b",
            "bun install",
            "cli/pnpm-i",
            "pnpm i → bun install"
        ),
        m!(
            r"\bpnpm\s+add\b",
            "bun add",
            "cli/pnpm-add",
            "pnpm add → bun add"
        ),
        m!(
            r"\bpnpm\s+remove\b",
            "bun remove",
            "cli/pnpm-remove",
            "pnpm remove → bun remove"
        ),
        m!(
            r"\bpnpm\s+rm\b",
            "bun remove",
            "cli/pnpm-rm",
            "pnpm rm → bun remove"
        ),
        m!(
            r"\bpnpm\s+run\s+",
            "bun run ",
            "cli/pnpm-run",
            "pnpm run → bun run"
        ),
        m!(
            r"\bpnpm\s+test\b",
            "bun test",
            "cli/pnpm-test",
            "pnpm test → bun test"
        ),
        m!(
            r"\bpnpm\s+start\b",
            "bun start",
            "cli/pnpm-start",
            "pnpm start → bun start"
        ),
        m!(
            r"\bpnpm\s+update\b",
            "bun update",
            "cli/pnpm-update",
            "pnpm update → bun update"
        ),
        m!(
            r"\bpnpm\s+dlx\s+",
            "bunx ",
            "cli/pnpm-dlx",
            "pnpm dlx → bunx"
        ),
        m!(
            r"\bpnpm\s+exec\s+",
            "bunx ",
            "cli/pnpm-exec",
            "pnpm exec → bunx"
        ),
        // yarn
        m!(
            r"\byarn\s+install\s+--frozen-lockfile\b",
            "bun install --frozen-lockfile",
            "cli/yarn-install-frozen",
            "yarn install --frozen-lockfile → bun install --frozen-lockfile"
        ),
        m!(
            r"\byarn\s+add\s+--dev\b",
            "bun add --dev",
            "cli/yarn-add-dev",
            "yarn add --dev → bun add --dev"
        ),
        m!(
            r"\byarn\s+add\s+-D\b",
            "bun add --dev",
            "cli/yarn-add-D",
            "yarn add -D → bun add --dev"
        ),
        m!(
            r"\byarn\s+add\b",
            "bun add",
            "cli/yarn-add",
            "yarn add → bun add"
        ),
        m!(
            r"\byarn\s+remove\b",
            "bun remove",
            "cli/yarn-remove",
            "yarn remove → bun remove"
        ),
        m!(
            r"\byarn\s+install\b",
            "bun install",
            "cli/yarn-install",
            "yarn install → bun install"
        ),
        m!(
            r"\byarn\s+run\s+",
            "bun run ",
            "cli/yarn-run",
            "yarn run → bun run"
        ),
        m!(
            r"\byarn\s+test\b",
            "bun test",
            "cli/yarn-test",
            "yarn test → bun test"
        ),
        m!(
            r"\byarn\s+start\b",
            "bun start",
            "cli/yarn-start",
            "yarn start → bun start"
        ),
        m!(
            r"\byarn\s+dlx\s+",
            "bunx ",
            "cli/yarn-dlx",
            "yarn dlx → bunx"
        ),
        // bare yarn at start of line → bun install (preserve leading ws)
        m!(
            r"(?m)^(\s*)yarn(\s|$)",
            "${1}bun install${2}",
            "cli/yarn-bare",
            "bare `yarn` → bun install"
        ),
    ]
});

static COMMENT_PREFIX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^\s*(#|//)").expect("invariant: COMMENT_PREFIX regex literal is valid")
});

/// Reproduit applyCliRules : applique chaque mapping en séquence, en
/// ignorant les lignes commentées, et collecte findings + contenu modifié.
pub fn apply_cli_rules(path: &str, source: &str) -> (Vec<Finding>, String) {
    let mut out = source.to_string();
    let mut findings: Vec<Finding> = Vec::new();
    // line_offsets est recalculé seulement si le contenu change entre règles
    // (c'est rare : la plupart des règles ne matchent pas). Mémoïsation simple.
    let mut offsets = line_offsets(&out);
    let mut offsets_stale = false;

    for rule in MAPPINGS.iter() {
        if offsets_stale {
            offsets = line_offsets(&out);
            offsets_stale = false;
        }
        let hits: Vec<(usize, String)> = rule
            .re
            .find_iter(&out)
            .filter(|mat| {
                let line_start = out[..mat.start()].rfind('\n').map(|p| p + 1).unwrap_or(0);
                let line_end = out[mat.start()..]
                    .find('\n')
                    .map(|p| mat.start() + p)
                    .unwrap_or(out.len());
                let line = &out[line_start..line_end];
                !COMMENT_PREFIX.is_match(line)
            })
            .map(|mat| (mat.start(), mat.as_str().to_string()))
            .collect();

        if hits.is_empty() {
            continue;
        }

        for (idx, text) in &hits {
            // Calcule la valeur de remplacement exacte en rejouant la regex
            // sur le hit pour conserver les back-refs ($1, $2).
            let rewritten = rule.re.replace(text, rule.replace).to_string();
            findings.push(make_finding(
                path,
                &offsets,
                *idx,
                rule.rule_id,
                rule.message,
                text.clone(),
                Some(rewritten),
                MakeFindingOpts {
                    autofix: Some(true),
                    ..Default::default()
                },
            ));
        }

        // Remplacement global (ignore les commentaires côté dommage : on
        // reproduit le comportement historique qui lui remplaçait globalement).
        let before = out.clone();
        out = rule.re.replace_all(&out, rule.replace).into_owned();
        if out != before {
            offsets_stale = true;
        }
    }

    (findings, out)
}
